File: prismpath/connector.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""prismpath.connector — The Connector SDK for PrismPath.

A subclassable developer API for building domain connectors against all **six hexagonal
ports** (adapters/ADAPTER_GUIDE.md): **Ingestion**, **Retrieval**, **Adjudicator**,
**Action/Sink**, **Attestation**, and **Deferral** — plus the node-handler dispatch that turns
a connector instance into a flow agent, and `PayloadFlattener`, the schema-flattening
middleware (nested API responses -> the FLAT key/value surface `when` predicates and guided
decoding actually work over; see the compliance adapter's "keep it FLAT" lesson).

Registry glue: `get_workers()` returns the exact `{name: callable(node, instruction, state)}`
shape the plugin registry consumes, so a pip-installable plugin module is one line —
``WORKERS = MyConnector().get_workers()`` — and every node bound with `@worker(plugin.name)`
carries `_worker` provenance in the transcript.
"""
from __future__ import annotations

import hashlib
import json
import inspect
import os
import re
from abc import ABC
from typing import Any, Dict, List, Callable, Optional, Union, Tuple

# Decorator to register node handlers
def node(name: str):
    """Decorator to mark a connector method as a node handler."""
    def decorator(func):
        func._prismpath_node = name
        return func
    return decorator

class BaseConnector(ABC):
    """
    Base class for all PrismPath Connectors.
    Abstracts the six ports (Ingestion, Retrieval, Adjudicator, Action/Sink, Attestation,
    Deferral) and simplifies node handler registration and agent dispatching.
    """
    def __init__(self, name: str, version: str = "1.0.0", deferral_store=None):
        self.name = name
        self.version = version
        self._handlers: Dict[str, Callable] = {}
        self._deferrals = deferral_store         # lazy default — see the `deferrals` property

        # Automatically discover methods decorated with @node
        for attr_name in dir(self):
            attr = getattr(self, attr_name)
            if hasattr(attr, "_prismpath_node"):
                node_name = getattr(attr, "_prismpath_node")
                self._handlers[node_name] = attr

    def register_handler(self, node_name: str, handler: Callable):
        """Programmatic registration of a handler function for a node."""
        self._handlers[node_name] = handler

    def handler(self, node_name: str):
        """Decorator to register a handler function for a node."""
        def decorator(func: Callable):
            self.register_handler(node_name, func)
            return func
        return decorator

    def __call__(self, node: str, instruction: str, state: dict) -> Any:
        """Allows the connector instance to act directly as an agent callable."""
        return self.agent(node, instruction, state)

    def agent(self, node: str, instruction: str, state: dict) -> Any:
        """
        Dispatches node execution to the registered handlers.
        Wraps executions to inject worker metadata and ensure standard compliance.
        """
        handler = self._handlers.get(node)
        if handler is None:
            return self.fallback(node, instruction, state)
        
        sig = inspect.signature(handler)
        params = list(sig.parameters.keys())
        num_args = len(params)
        
        if num_args == 1:
            res = handler(state)
        elif num_args == 2:
            res = handler(instruction, state)
        else:
            res = handler(node, instruction, state)
            
        if isinstance(res, dict):
            res.setdefault("_worker", f"{self.name}.{node}")
        return res

    def fallback(self, node: str, instruction: str, state: dict) -> Any:
        """Fallback behavior when no handler is registered for the node."""
        raise NotImplementedError(
            f"Node '{node}' is not implemented in connector '{self.name}'"
        )

    def get_workers(self) -> Dict[str, Callable[[str, str, dict], Any]]:
        """
        Returns a dictionary of workers compatible with the prismpath plugins registry.
        """
        return {
            node_name: lambda n, inst, s, name=node_name: self.agent(name, inst, s)
            for node_name in self._handlers
        }

    # --- INGESTION PORT ---
    def ingest_payload(self, raw_data: Any) -> Dict[str, Any]:
        """Override to process, sanitize, or parse raw incoming data."""
        if isinstance(raw_data, dict):
            return raw_data
        return {"raw": raw_data}

    def compute_ingestion_hash(self, data: Dict[str, Any]) -> str:
        """Computes a content-addressable hash for ingestion payloads."""
        body = json.dumps(data, sort_keys=True).encode()
        return "sha256:" + hashlib.sha256(body).hexdigest()[:16]

    # --- RETRIEVAL PORT ---
    def retrieve_criteria(self, query: str) -> Any:
        """Override to fetch domain knowledge or catalog criteria."""
        return None

    def compute_knowledge_hash(self, kb_data: Any) -> str:
        """Computes a content-addressable hash for knowledge base / catalog data."""
        body = json.dumps(kb_data, sort_keys=True).encode()
        return "sha256:" + hashlib.sha256(body).hexdigest()[:16]

    # --- ADJUDICATOR PORT ---
    def adjudication_prompt(self, payload: Dict[str, Any], criteria: Any = None,
                            schema: Optional[Dict[str, Any]] = None) -> str:
        """The default prompt surface: the payload FLATTENED to key/value lines (the
        schema-flattening middleware in action — nested objects destabilize guided decoding),
        optional retrieved criteria, and a flat-JSON reply instruction when a schema is given.
        Override to shape domain prompts; the flat discipline is the part worth keeping."""
        flat = PayloadFlattener().flatten(payload) if isinstance(payload, (dict, list)) \
            else {"input": payload}
        lines = [f"{k}: {v}" for k, v in sorted(flat.items())]
        parts = ["\n".join(lines)]
        if criteria is not None:
            parts.append(f"CRITERIA:\n{criteria}")
        if schema:
            keys = ", ".join(sorted((schema.get("properties") or schema).keys()))
            parts.append(f"Reply with ONE flat JSON object (no nesting) with keys: {keys}.")
        return "\n\n".join(parts)

    def adjudicate(self, payload: Dict[str, Any], generate: Optional[Callable[[str], str]] = None,
                   schema: Optional[Dict[str, Any]] = None, guard=None,
                   criteria: Any = None) -> Dict[str, Any]:
        """Run one adjudication through the port. `generate` is ANY text->text callable — a
        served model, a local pipeline, a comparator bank, a human console; the port never
        assumes an LLM (the adapter contracts' rule — a hardware target can drive this port
        with comparators). When a `guard` (prismpath.guard.Guard) is given, the exchange crosses
        `guarded_exchange`, so denied input never reaches the model and denied output never
        returns. The reply's first JSON object becomes the outcome dict (fields for `when`
        predicates); a non-JSON reply degrades to {"text": reply}."""
        if generate is None:
            raise TypeError(
                "adjudicate() needs a `generate` callable (text -> text): the Adjudicator port "
                "does not assume an LLM — pass your model client, comparator, or reviewer bridge")
        prompt = self.adjudication_prompt(payload, criteria=criteria, schema=schema)
        if guard is not None:
            from prismpath.guard import guarded_exchange
            reply = guarded_exchange(guard, prompt, generate)
        else:
            reply = generate(prompt)
        m = re.search(r"\{.*\}", reply or "", re.S)
        if m:
            try:
                out = json.loads(m.group(0))
                if isinstance(out, dict):
                    out.setdefault("text", (reply or "").strip())
                    return out
            except ValueError:
                pass
        return {"text": (reply or "").strip()}

    # --- ACTION / SINK PORT ---
    def emit_record(self, result: Dict[str, Any], destination: str, key: str = "id") -> Any:
        """Default sink: idempotent JSONL append (`ledger_runner.upsert_jsonl`) keyed on `key`,
        so a replayed item never double-writes — the property `run_ledgered_loop` relies on.
        Records missing the key field append unconditionally. Override for OSCAL/CycloneDX/
        webhook emitters."""
        from prismpath.ledger_runner import upsert_jsonl
        os.makedirs(os.path.dirname(destination) or ".", exist_ok=True)
        if result.get(key) is None:
            with open(destination, "a") as f:
                f.write(json.dumps(result, sort_keys=True) + "\n")
            return True
        return upsert_jsonl(destination, result, key=key)

    # --- ATTESTATION PORT ---
    @staticmethod
    def policy_hash_for(flow_path: str) -> str:
        """The policy hash `attest_decision` should bind: the content hash of the governing
        flow document (`checkpoint.flow_hash`) — a decision provably tied to the exact flow
        version that made it."""
        from prismpath.checkpoint import flow_hash
        return flow_hash(flow_path)

    def attest_decision(
        self,
        outcome: Dict[str, Any],
        policy_hash: str,
        gate_id: str,
        ingestion_hashes: List[str],
        kb_hash: str,
        label: Optional[str] = None
    ) -> Dict[str, Any]:
        """Default core attestation binding using ledger_airgap. Pass
        `policy_hash_for(flow_path)` as `policy_hash` to bind the governing document."""
        from prismpath import ledger_airgap
        root_hex = hashlib.sha256(json.dumps(outcome, sort_keys=True).encode()).hexdigest()
        return ledger_airgap.provenance_manifest(
            root_hex=root_hex,
            label=label or f"{self.name}:decision",
            policy_hash=policy_hash,
            gate_id=gate_id,
            ingestion_hashes=ingestion_hashes,
            knowledge_base_hash=kb_hash
        )

    # --- DEFERRAL PORT ---
    @property
    def deferrals(self):
        """The connector's DeferralStore (the port base in `prismpath.deferral`). Defaults to a
        `FileDeferralStore` under `./<name>.deferrals` on first use; inject any backend via
        `__init__(deferral_store=...)`."""
        if self._deferrals is None:
            from prismpath.deferral import FileDeferralStore
            self._deferrals = FileDeferralStore(f"{self.name}.deferrals")
        return self._deferrals

    def defer_decision(self, unit_id: str, reason: str, state: Dict[str, Any],
                       prior_output: Any = None):
        """Suspend a unit for human review / evidence discovery. One primitive serves both the
        HITL-override and evidence-request loops; `resume_decision` closes it with the actor
        recorded."""
        return self.deferrals.defer(unit_id, reason, state, prior_output=prior_output)

    def resume_decision(self, unit_id: str, resolution: Dict[str, Any], actor: str):
        return self.deferrals.resume(unit_id, resolution, actor)

    def pending_deferrals(self):
        return self.deferrals.pending()


class PayloadFlattener:
    """
    Utility class to flatten nested dictionary payloads and apply custom mapping rules.
    Helps connectors format nested API responses into flat key-value pairs for the LLM.
    """
    def __init__(self, delimiter: str = "."):
        self.delimiter = delimiter

    def flatten(self, data: Any, prefix: str = "") -> Dict[str, Any]:
        """
        Recursively flattens a nested dict/list structure into a flat dict.
        Lists are represented as index-based keys (e.g., 'roles.0') and also
        joined as comma-separated strings at their parent prefix.
        """
        flat = {}
        if isinstance(data, dict):
            for k, v in data.items():
                new_key = f"{prefix}{k}" if not prefix else f"{prefix}{self.delimiter}{k}"
                flat.update(self.flatten(v, new_key))
        elif isinstance(data, list):
            if prefix:
                flat[prefix] = ", ".join(str(item) for item in data if not isinstance(item, (dict, list)))
            for idx, item in enumerate(data):
                new_key = f"{prefix}{self.delimiter}{idx}" if prefix else str(idx)
                flat.update(self.flatten(item, new_key))
        else:
            if prefix:
                flat[prefix] = data
        return flat

    def map_fields(self, data: Any, rules: Dict[str, Union[str, Tuple[str, Callable]]]) -> Dict[str, Any]:
        """
        Maps fields from a nested structure using rules.
        Rules definition:
        {target_key: source_path}
        or
        {target_key: (source_path, transform_fn)}
        """
        mapped = {}
        # Pre-flatten only if fallback is needed, but we can do it lazily
        flat_data = None
        for target, rule in rules.items():
            if isinstance(rule, tuple):
                source_path, transform = rule
            else:
                source_path, transform = rule, None
            
            # Resolve from raw nested structure first to preserve original types for transformers
            val = self._resolve_path(data, source_path)
            if val is None:
                if flat_data is None:
                    flat_data = self.flatten(data)
                val = flat_data.get(source_path)
                
            if val is not None and transform is not None:
                try:
                    val = transform(val)
                except Exception:
                    pass
            mapped[target] = val
        return mapped


    def _resolve_path(self, data: Any, path: str) -> Any:
        """Resolves a nested path manually in the original raw data structure."""
        parts = path.split(self.delimiter)
        current = data
        for part in parts:
            if isinstance(current, dict):
                current = current.get(part)
            elif isinstance(current, list):
                try:
                    current = current[int(part)]
                except (ValueError, IndexError):
                    return None
            else:
                return None
        return current

    @staticmethod
    def to_delimited_string(key: str, delimiter: str = ", ") -> Callable[[List[Dict[str, Any]]], str]:
        """Extracts a specific key from a list of dicts and joins them into a string."""
        def transformer(items: List[Dict[str, Any]]) -> str:
            if not isinstance(items, list):
                return str(items)
            extracted = []
            for item in items:
                if isinstance(item, dict):
                    val = item.get(key)
                    if val is not None:
                        extracted.append(str(val))
                else:
                    extracted.append(str(item))
            return delimiter.join(extracted)
        return transformer

    @staticmethod
    def format_datetime(output_format: str = "%Y-%m-%d") -> Callable[[str], str]:
        """Parses a datetime string and formats it to output_format."""
        from datetime import datetime
        def transformer(val: str) -> str:
            formats = [
                "%Y-%m-%dT%H:%M:%S.%fZ",
                "%Y-%m-%dT%H:%M:%SZ",
                "%Y-%m-%d %H:%M:%S",
                "%Y-%m-%d"
            ]
            for fmt in formats:
                try:
                    dt = datetime.strptime(val, fmt)
                    return dt.strftime(output_format)
                except ValueError:
                    continue
            return val
        return transformer


class SystemTelemetry(BaseConnector):
    """
    Ingestion Connector to dynamically probe system hardware telemetry (RAM, VRAM, CPU cores)
    on launch.
    """
    def __init__(self):
        super().__init__("SystemTelemetry", "1.0.0")

    def ingest_payload(self, raw_data: Any = None) -> Dict[str, Any]:
        """Probes system hardware: RAM, VRAM, CPU cores."""
        import os

        # Query CPU cores
        try:
            cpu_cores = os.cpu_count() or 1
        except Exception:
            cpu_cores = 1

        # Query System RAM (MB to GB)
        ram_gb = 8.0
        try:
            if os.path.exists("/proc/meminfo"):
                with open("/proc/meminfo", "r") as f:
                    for line in f:
                        if line.startswith("MemTotal:"):
                            parts = line.split()
                            if len(parts) >= 2:
                                ram_gb = round(float(parts[1]) / (1024 * 1024), 2)
                                break
            else:
                import psutil
                ram_gb = round(psutil.virtual_memory().total / (1024 ** 3), 2)
        except Exception:
            ram_gb = 8.0

        # Query VRAM (MB to GB via nvidia-smi if available)
        vram_gb = 0.0
        try:
            import subprocess
            res = subprocess.run(
                ["nvidia-smi", "--query-gpu=memory.total", "--format=csv,noheader,nounits"],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                check=True
            )
            val = float(res.stdout.strip())
            vram_gb = round(val / 1024.0, 2)
        except Exception:
            vram_gb = 0.0

        return {
            "ram_gb": ram_gb,
            "vram_gb": vram_gb,
            "cpu_cores": cpu_cores,
            "ram_source": "host",
            "unified": False
        }

File: prismpath-rs/src/connector.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! connector.rs: the six-port Connector SDK, natively (feature `durable`).
//!
//! Faithful port of the PORT SURFACE of `prismpath/connector.py`: Ingestion, Retrieval,
//! Adjudicator, Action/Sink, Attestation, Deferral. Domains implement the trait; the defaults
//! reproduce the reference's portable behaviors bit-for-bit where they are data (hashes, the
//! flattened prompt surface, the attestation manifest: gated against `conformance/connector.json`)
//! and structurally where they are I/O (the idempotent JSONL sink). The Adjudicator port takes ANY
//! text->text callable: a served model, a comparator bank, a human console: never assuming an
//! LLM, exactly like the reference.
//!
//! Deliberately minimal vs Python: the guard hook (content safety is delegated by design: see
//! prismpath-rs/CONFORMANCE.md's scope boundary). The deferral port ships BOTH stores: in-memory
//! and `FileDeferralStore`, byte-compatible with the reference's directory layout so either
//! runtime can list and resume units the other deferred.

use crate::durable::{flow_hash, provenance_manifest, py_canonical_string};
use crate::{py_str, Value as EngineValue};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::io::Write;

/// `PayloadFlattener.flatten`: nested dict/list -> flat key/value strings. Lists appear BOTH as a
/// comma-joined scalar string at their parent prefix and as per-index keys.
pub fn flatten(data: &Value, prefix: &str, delimiter: &str, out: &mut BTreeMap<String, String>) {
    match data {
        Value::Object(obj_val) => {
            for (key_str, val) in obj_val {
                let key = if prefix.is_empty() {
                    key_str.clone()
                } else {
                    format!("{prefix}{delimiter}{key_str}")
                };
                flatten(val, &key, delimiter, out);
            }
        }
        Value::Array(arr_val) => {
            if !prefix.is_empty() {
                let joined: Vec<String> = arr_val
                    .iter()
                    .filter(|item| !item.is_object() && !item.is_array())
                    .map(|item| py_str(&EngineValue::from_json(item)))
                    .collect();
                out.insert(prefix.to_string(), joined.join(", "));
            }
            for (idx, item) in arr_val.iter().enumerate() {
                let key = if prefix.is_empty() {
                    idx.to_string()
                } else {
                    format!("{prefix}{delimiter}{idx}")
                };
                flatten(item, &key, delimiter, out);
            }
        }
        scalar => {
            if !prefix.is_empty() {
                out.insert(prefix.to_string(), py_str(&EngineValue::from_json(scalar)));
            }
        }
    }
}

/// The reference's short content hash: "sha256:" + first 16 hex chars over the spaced canonical
/// JSON (`json.dumps(sort_keys=True)`).
pub fn short_hash(data: &Value) -> String {
    let digest = hex::encode(Sha256::digest(py_canonical_string(data, true).as_bytes()));
    format!("sha256:{}", &digest[..16])
}

/// The six-port Connector. Every method has the reference's default; override per domain.
pub trait Connector {
    fn name(&self) -> &str;

    // --- INGESTION PORT ---
    fn ingest_payload(&self, raw: &Value) -> Value {
        if raw.is_object() {
            raw.clone()
        } else {
            json!({"raw": raw})
        }
    }
    fn ingestion_hash(&self, data: &Value) -> String {
        short_hash(data)
    }

    // --- RETRIEVAL PORT ---
    fn retrieve_criteria(&self, _query: &str) -> Option<Value> {
        None
    }
    fn knowledge_hash(&self, kb: &Value) -> String {
        short_hash(kb)
    }

    // --- ADJUDICATOR PORT ---
    /// The default prompt surface: the payload FLATTENED to sorted key/value lines, optional
    /// criteria, and a flat-JSON reply instruction when a schema is given.
    fn adjudication_prompt(&self, payload: &Value, criteria: Option<&str>, schema: Option<&Value>) -> String {
        let mut flat = BTreeMap::new();
        if payload.is_object() || payload.is_array() {
            flatten(payload, "", ".", &mut flat);
        } else {
            flat.insert("input".to_string(), py_str(&EngineValue::from_json(payload)));
        }
        let lines: Vec<String> = flat.iter().map(|(key, val)| format!("{key}: {val}")).collect();
        let mut parts = vec![lines.join("\n")];
        if let Some(crit) = criteria {
            parts.push(format!("CRITERIA:\n{crit}"));
        }
        if let Some(schema_val) = schema {
            let props = schema_val.get("properties").unwrap_or(schema_val);
            let mut keys: Vec<&String> = props.as_object().map(|obj_val| obj_val.keys().collect()).unwrap_or_default();
            keys.sort();
            let joined = keys.iter().map(|str_val| str_val.as_str()).collect::<Vec<_>>().join(", ");
            parts.push(format!("Reply with ONE flat JSON object (no nesting) with keys: {joined}."));
        }
        parts.join("\n\n")
    }

    /// Run one adjudication: `generate` is ANY text->text callable. The reply's first JSON
    /// object becomes the outcome dict; a non-JSON reply degrades to {"text": reply}.
    fn adjudicate(
        &self,
        payload: &Value,
        generate: &mut dyn FnMut(&str) -> String,
        criteria: Option<&str>,
        schema: Option<&Value>,
    ) -> Value {
        let prompt = self.adjudication_prompt(payload, criteria, schema);
        let reply = generate(&prompt);
        if let (Some(start_idx), Some(end_idx)) = (reply.find('{'), reply.rfind('}')) {
            if start_idx < end_idx {
                if let Ok(Value::Object(mut out_map)) =
                    serde_json::from_str::<Value>(&reply[start_idx..=end_idx])
                {
                    out_map.entry("text".to_string())
                        .or_insert_with(|| Value::String(reply.trim().to_string()));
                    return Value::Object(out_map);
                }
            }
        }
        json!({"text": reply.trim()})
    }

    // --- ACTION / SINK PORT ---
    /// Idempotent JSONL append keyed on `key`: a replayed item never double-writes.
    fn emit_record(&self, result: &Value, destination: &str, key: &str) -> Result<bool, String> {
        if let Some(dir_path) = std::path::Path::new(destination).parent() {
            if !dir_path.as_os_str().is_empty() {
                std::fs::create_dir_all(dir_path).map_err(|err| err.to_string())?;
            }
        }
        let key_val = result.get(key);
        if key_val.is_none() || key_val == Some(&Value::Null) {
            let mut file_handle = std::fs::OpenOptions::new()
                .create(true)
                .append(true)
                .open(destination)
                .map_err(|err| err.to_string())?;
            writeln!(file_handle, "{}", py_canonical_string(result, true)).map_err(|err| err.to_string())?;
            return Ok(true);
        }
        // upsert: replace the line whose `key` matches, else append
        let existing = std::fs::read_to_string(destination).unwrap_or_default();
        let mut lines: Vec<String> = Vec::new();
        let mut replaced = false;
        for line_str in existing.lines().filter(|line_item| !line_item.trim().is_empty()) {
            match serde_json::from_str::<Value>(line_str) {
                Ok(row) if row.get(key) == key_val => {
                    lines.push(py_canonical_string(result, true));
                    replaced = true;
                }
                _ => lines.push(line_str.to_string()),
            }
        }
        if !replaced {
            lines.push(py_canonical_string(result, true));
        }
        std::fs::write(destination, lines.join("\n") + "\n").map_err(|err| err.to_string())?;
        Ok(true)
    }

    // --- ATTESTATION PORT ---
    /// The policy hash to bind: the content hash of the governing flow document.
    fn policy_hash_for(&self, flow_path: &str) -> String {
        flow_hash(flow_path)
    }

    /// Core attestation binding: outcome root + provenance manifest (durable::provenance_manifest,
    /// C1). `created` injected for determinism, as in the durable layer.
    #[allow(clippy::too_many_arguments)]
    fn attest_decision(
        &self,
        outcome: &Value,
        policy_hash: &str,
        gate_id: &str,
        ingestion_hashes: &[&str],
        kb_hash: &str,
        label: Option<&str>,
        created: &str,
    ) -> Value {
        let root = hex::encode(Sha256::digest(py_canonical_string(outcome, true).as_bytes()));
        let default_label = format!("{}:decision", self.name());
        provenance_manifest(
            &root,
            label.unwrap_or(&default_label),
            created,
            Some(policy_hash),
            Some(gate_id),
            ingestion_hashes,
            Some(kb_hash),
        )
    }
}

/// Reference no-op connector: echoes payloads through the default ports (the SDK's smoke surface).
pub struct EchoConnector {
    pub name: String,
}
impl Connector for EchoConnector {
    fn name(&self) -> &str {
        &self.name
    }
}

// --- DEFERRAL PORT ---

#[derive(Debug, Clone)]
pub struct Deferral {
    pub unit_id: String,
    pub reason: String,
    pub state: Value,
    pub prior_output: Option<Value>,
    pub resolution: Option<Value>,
    pub actor: Option<String>,
}

pub trait DeferralStore {
    fn defer(&mut self, unit_id: &str, reason: &str, state: Value, prior_output: Option<Value>);
    fn pending(&self) -> Vec<&Deferral>;
    fn resume(&mut self, unit_id: &str, resolution: Value, actor: &str) -> bool;
}

/// File-backed deferral store, BYTE-COMPATIBLE with the reference's `FileDeferralStore`
/// (`prismpath/deferral.py`): one JSON record per unit at
/// `<dir>/<sha256(unit_id)[:16]>.json` with the same field set and lifecycle
/// (`status: pending -> resolved`, ISO-8601Z timestamps) - so a unit deferred by either runtime
/// can be listed and resumed by the other against one shared directory.
pub struct FileDeferralStore {
    pub dir: std::path::PathBuf,
}

impl FileDeferralStore {
    pub fn new(dir: impl Into<std::path::PathBuf>) -> std::io::Result<FileDeferralStore> {
        let dir = dir.into();
        std::fs::create_dir_all(&dir)?;
        Ok(FileDeferralStore { dir })
    }

    fn path(&self, unit_id: &str) -> std::path::PathBuf {
        let hash_str = hex::encode(Sha256::digest(unit_id.as_bytes()));
        self.dir.join(format!("{}.json", &hash_str[..16]))
    }

    fn now_iso() -> String {
        let secs = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|dur| dur.as_secs())
            .unwrap_or(0);
        let (days, hours, mins, seconds) = (secs / 86_400, (secs % 86_400) / 3600, (secs % 3600) / 60, secs % 60);
        let z_val = days as i64 + 719_468;
        let era = z_val.div_euclid(146_097);
        let doe = z_val.rem_euclid(146_097);
        let yoe = (doe - doe / 1460 + doe / 36_524 - doe / 146_096) / 365;
        let year_val = yoe + era * 400;
        let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
        let month_pos = (5 * doy + 2) / 153;
        let day_val = doy - (153 * month_pos + 2) / 5 + 1;
        let month_val = if month_pos < 10 { month_pos + 3 } else { month_pos - 9 };
        let final_year = if month_val <= 2 { year_val + 1 } else { year_val };
        format!("{final_year:04}-{month_val:02}-{day_val:02}T{hours:02}:{mins:02}:{seconds:02}Z")
    }

    pub fn defer(
        &self,
        unit_id: &str,
        reason: &str,
        state: Value,
        prior_output: Option<Value>,
    ) -> Result<Value, String> {
        let rec = json!({
            "unit_id": unit_id, "reason": reason, "state": state,
            "prior_output": prior_output.unwrap_or(Value::Null),
            "status": "pending", "deferred_at": Self::now_iso(),
            "resolution": Value::Null, "actor": Value::Null, "resolved_at": Value::Null,
        });
        std::fs::write(
            self.path(unit_id),
            serde_json::to_string_pretty(&rec).map_err(|err| err.to_string())?,
        )
        .map_err(|err| err.to_string())?;
        Ok(rec)
    }

    pub fn get(&self, unit_id: &str) -> Option<Value> {
        std::fs::read_to_string(self.path(unit_id))
            .ok()
            .and_then(|text| serde_json::from_str(&text).ok())
    }

    pub fn pending(&self) -> Vec<Value> {
        let mut names: Vec<_> = std::fs::read_dir(&self.dir)
            .map(|read_dir| {
                read_dir.filter_map(|entry_res| entry_res.ok())
                    .map(|entry| entry.file_name().to_string_lossy().into_owned())
                    .filter(|name_str| name_str.ends_with(".json"))
                    .collect()
            })
            .unwrap_or_default();
        names.sort();
        names
            .into_iter()
            .filter_map(|name_str| {
                let text = std::fs::read_to_string(self.dir.join(name_str)).ok()?;
                let rec: Value = serde_json::from_str(&text).ok()?;
                (rec.get("status").and_then(|status_val| status_val.as_str()) == Some("pending")).then_some(rec)
            })
            .collect()
    }

    pub fn resume(&self, unit_id: &str, resolution: Value, actor: &str) -> Result<Value, String> {
        let mut rec = self.get(unit_id).ok_or(format!("no deferred unit {unit_id}"))?;
        let status = rec.get("status").and_then(|status_val| status_val.as_str()).unwrap_or("");
        if status != "pending" {
            return Err(format!("unit {unit_id} already {status}"));
        }
        rec["status"] = json!("resolved");
        rec["resolution"] = resolution;
        rec["actor"] = json!(actor);
        rec["resolved_at"] = json!(Self::now_iso());
        std::fs::write(
            self.path(unit_id),
            serde_json::to_string_pretty(&rec).map_err(|err| err.to_string())?,
        )
        .map_err(|err| err.to_string())?;
        Ok(rec)
    }
}

#[derive(Default)]
pub struct MemDeferralStore {
    pub items: Vec<Deferral>,
}
impl DeferralStore for MemDeferralStore {
    fn defer(&mut self, unit_id: &str, reason: &str, state: Value, prior_output: Option<Value>) {
        self.items.push(Deferral {
            unit_id: unit_id.to_string(),
            reason: reason.to_string(),
            state,
            prior_output,
            resolution: None,
            actor: None,
        });
    }
    fn pending(&self) -> Vec<&Deferral> {
        self.items.iter().filter(|item| item.resolution.is_none()).collect()
    }
    fn resume(&mut self, unit_id: &str, resolution: Value, actor: &str) -> bool {
        for item in self.items.iter_mut() {
            if item.unit_id == unit_id && item.resolution.is_none() {
                item.resolution = Some(resolution);
                item.actor = Some(actor.to_string());
                return true;
            }
        }
        false
    }
}
